hibernate-core/src/main/java/org/hibernate/query/hql/internal/SemanticQueryBuilder.java

```diff
@@ -301,42 +303,81 @@ public SqmSelectStatement visitSelectStatement(HqlParser.SelectStatementContext
 	}
 
 	@Override
-	public SqmInsertSelectStatement visitInsertStatement(HqlParser.InsertStatementContext ctx) {
-		final EntityDomainType<?> targetType = visitEntityName( ctx.insertSpec().intoSpec().entityName() );
+	public SqmInsertStatement visitInsertStatement(HqlParser.InsertStatementContext ctx) {
 
-		final SqmRoot<?> root = new SqmRoot<>( targetType, null, creationContext.getNodeBuilder() );
-		processingStateStack.getCurrent().getPathRegistry().register( root );
-
-		// for now we only support the INSERT-SELECT form
-		final SqmInsertSelectStatement insertStatement = new SqmInsertSelectStatement<>(
-				root,
-				creationContext.getQueryEngine().getCriteriaBuilder()
+		final SqmRoot<?> root = new SqmRoot<>(
+				visitEntityName( ctx.rootEntity().entityName() ),
+				visitIdentificationVariableDef( ctx.rootEntity().identificationVariableDef() ),
+				creationContext.getNodeBuilder()
 		);
-		parameterCollector = insertStatement;
 
-		processingStateStack.push( new SqmDmlCreationProcessingState( insertStatement, this ) );
+		if ( ctx.querySpec()!=null ) {
+			final SqmInsertSelectStatement<?> insertStatement = new SqmInsertSelectStatement<>( root, creationContext.getNodeBuilder() );
+			parameterCollector = insertStatement;
+			final SqmDmlCreationProcessingState processingState = new SqmDmlCreationProcessingState(
+					insertStatement,
+					this
+			);
+
+			processingStateStack.push( processingState );
+			processingState.getPathRegistry().register( root );
```

Comment:
is it better to move the last statement into the try block below so we can ensure the `processingState` is always popped even when the registering process throws some exception (I checked its internal implementation and found it seems possible)?
ignore if irrelevant. Just a little bit confusing from casual code reader's eyes.

Reply:
Perhaps, but that's not how it was written before this commit, and it's not what happens in `visitUpdateStatement()`, so it's not related to this PR. Anyway, If `register()` throws, then AFAIK the whole query translation process is hosed, so I doubt the try/finally block is even really useful for anything.